pkg/storage/storagewrappers/boundedconcurrency.go

```diff
@@ -0,0 +1,86 @@
+package storagewrappers
+
+import (
+	"context"
+	"time"
+
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const TimeWaitingSpanAttribute = "time_waiting"
+
+var _ storage.RelationshipTupleReader = (*boundedConcurrencyTupleReader)(nil)
+
+var (
+	timeWaitingCounter = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name:    "time_waiting_for_bounded_query",
+		Help:    "Time (in ms) spent waiting for Read, ReadUserTuple and ReadUsersetTuples calls to the datastore",
+		Buckets: []float64{10, 25, 50, 100, 1000, 5000}, // milliseconds. Upper bound is config.UpstreamTimeout
+	})
+)
+
+type boundedConcurrencyTupleReader struct {
+	storage.RelationshipTupleReader
+	limiter chan struct{}
+}
+
+// NewBoundedConcurrencyTupleReader returns a wrapper over a datastore that makes sure that there are, at most,
+// N concurrent calls to Read, ReadUserTuple and ReadUsersetTuples.
+// Consumers can then rest assured that one client will not hoard all the database connections available.
+func NewBoundedConcurrencyTupleReader(wrapped storage.RelationshipTupleReader, N uint32) *boundedConcurrencyTupleReader {
```

Comment:
```suggestion
func NewBoundedConcurrencyTupleReader(wrapped storage.RelationshipTupleReader, concurrencyLimit uint32) *boundedConcurrencyTupleReader {
```

Reply:
I wrote `N` because I wanted the godoc to be clear that N was the parameter. There is no way of having nice godocs that reference parameters :(